src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1489,45 +1489,82 @@ private bool TryProcessCompilationCompleted(CompilationCompletedEvent endEvent,
+                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation, analyzerOptions);
                 if (filteredDiagnostic != null)
                 {
                     addDiagnosticCore(filteredDiagnostic);
                 }
             };
         }
 
-        internal static Action<Diagnostic, DiagnosticAnalyzer, bool> GetDiagnosticSink(Action<Diagnostic, DiagnosticAnalyzer, bool> addLocalDiagnosticCore, Compilation compilation)
+        internal static Action<Diagnostic, DiagnosticAnalyzer, bool> GetDiagnosticSink(Action<Diagnostic, DiagnosticAnalyzer, bool> addLocalDiagnosticCore, Compilation compilation, AnalyzerOptions analyzerOptions)
         {
             return (diagnostic, analyzer, isSyntaxDiagnostic) =>
             {
-                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation);
+                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation, analyzerOptions);
                 if (filteredDiagnostic != null)
                 {
                     addLocalDiagnosticCore(filteredDiagnostic, analyzer, isSyntaxDiagnostic);
                 }
             };
         }
 
-        internal static Action<Diagnostic, DiagnosticAnalyzer> GetDiagnosticSink(Action<Diagnostic, DiagnosticAnalyzer> addDiagnosticCore, Compilation compilation)
+        internal static Action<Diagnostic, DiagnosticAnalyzer> GetDiagnosticSink(Action<Diagnostic, DiagnosticAnalyzer> addDiagnosticCore, Compilation compilation, AnalyzerOptions analyzerOptions)
         {
             return (diagnostic, analyzer) =>
             {
-                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation);
+                var filteredDiagnostic = GetFilteredDiagnostic(diagnostic, compilation, analyzerOptions);
                 if (filteredDiagnostic != null)
                 {
                     addDiagnosticCore(filteredDiagnostic, analyzer);
                 }
             };
         }
 
-        private static Diagnostic GetFilteredDiagnostic(Diagnostic diagnostic, Compilation compilation)
+        private static Diagnostic GetFilteredDiagnostic(Diagnostic diagnostic, Compilation compilation, AnalyzerOptions analyzerOptions)
         {
-            return compilation.Options.FilterDiagnostic(diagnostic);
+            diagnostic = compilation.Options.FilterDiagnostic(diagnostic);
+
+            // Apply bulk configuration from analyzer options, if applicable.
+            var tree = diagnostic?.Location.SourceTree;
+            if (tree == null || analyzerOptions == null)
+            {
+                return diagnostic;
+            }
+
+            // If user has explicitly configured severity for this diagnostic ID, that should be respected and
+            // bulk configuration should not be applied.
+            // For example, 'dotnet_diagnostic.CA1000.severity = error'
+            if (compilation.Options.SpecificDiagnosticOptions.ContainsKey(diagnostic.Id) ||
```

Comment:
This doesn't look right. I think this is overriding `pragma`. We need to figure out some way to not re-implement the compiler logic here, since if one piece changes, this code will break.

Reply:
`CompilationOptions.FilterDiagnostic` will return `null` for suppressed diagnostic. We will bail out upfront for any suppression case.

Reply:
What about global diagnostic settings?

Reply:
Also, what if we add `#pragma warning enable`?

Reply:
Neither of those should affect how this feature should work - they do not show explicit user intent to configure severities of specific diagnostics, so we should respect the bulk configuration.

Reply:
Right, but they don't in this case, do they? If you have global diagnostic suppress, and someone has a diagnostic configured as warning, then the Filter will return `null`, and the category configuration will not be considered, right?

Reply:
> If you have global diagnostic suppress

Which compiler switch allows the global diagnostic option to be suppress? As in, can you point me to a repro case where I can see the behavior you mention?

Reply:
`Compilation.GeneralDiagnosticOption`

Reply:
Specifically, a diagnostic with default warning severity, where `GeneralDiagnosticOption` is `Suppress`.

See here for what I'm thinking of: https://github.com/dotnet/roslyn/blob/cf104c5204f51d75e79c93935b8bb96b9c8d3796/src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs#L227

Reply:
I don't believe any compiler switch allows to set that to suppress and I am not sure we have crisp definition of semantics associated with `Compilation.GeneralDiagnosticOption = ReportDiagnostic.Suppress`. IMO, it can mean couple of things:

1. Never ever report a non-compiler error diagnostic for me
2. Never ever report a non-compiler error diagnostic for me unless I explicitly have an ID specific configuration to unsuppress the rule

I don't think we want to apply bulk configuration and turn on huge bucket of diagnostics in either case.

Reply:
> I don't believe any compiler switch allows to set that to suppress

How is that relevant?

>  I am not sure we have crisp definition of semantics associated with Compilation.GeneralDiagnosticOption = ReportDiagnostic.Suppress

I think the principle I stated before, namely the more specific rules override the more general rules, say that category is more specific than universal, so category should override universal configuration.

Reply:
> How is that relevant?

It is relevant because there is no realistic user scenario where this is applicable. There is no way a user can reach this state either from command line builds or IDE live analysis.

Reply:
If you insist, I can file a tracking issue to enable bulk configuration functionality in presence of Global diagnostic option = suppress, which as of today can only happen for a custom analyzer host. If there is any user feedback that this affected them, we can decide to implement it.

Reply:
> It is relevant because there is no realistic user scenario where this is applicable. There is no way a user can reach this state either from command line builds or IDE live analysis.

That's not how this works. We ship an API as well as a host.

Reply:
> That's not how this works. We ship an API as well as a host.

This is a new analyzer feature aimed for real analyzer users. Why is it not acceptable to curtail a new feature to only work for subset of scenarios, especially given the scenario is not a mainline one? The only place where feature will not light up is if someone writes a custom analyzer host, and executes it on a compilation with global suppress option, which also has bulk configuration settings in editorconfig. I don't believe that is a compelling scenario to enable until we get *real* customer requests for this scenario.

Reply:
Changing this behavior later would be a breaking change for anyone using the API, since it would cause diagnostics that were previously suppressed with GeneralDiagnosticOption to respect `category`, which may be a higher diagnostic level than `suppress`

Reply:
> Changing this behavior later would be a breaking change for anyone using the API

It would be a side-effect of new functionality lighting up for a new feature (bulk configuring analyzer diagnostics) which some user needs to explicitly enable by adding such entries. Until then, the behavior is identical to user not having any bulk configuration entries.

Reply:
Talked to @jaredpar, he said he wasn't worried about fixing this later, so if you want to a file a bug we can handle it then.

Reply:
Thanks, filed https://github.com/dotnet/roslyn/issues/39033